libraries/chain/block_log.cpp

```diff
@@ -1475,4 +1503,240 @@ namespace eosio { namespace chain {
    bool block_log::exists(const fc::path& data_dir) {
       return fc::exists(data_dir / "blocks.log") && fc::exists(data_dir / "blocks.index");
    }
+
+   // For applications like trim from last, this function is used to
+   // make block and index files match each other.
+   // If head block in log file is invalid, search for the last 
+   // good block and discard the rest.
+   // If index file does not agree with block file, trim or pad index file
+   // when necessary.
+   void trim_data::check_blockfile_indexfile() {
+      if (ind_in == nullptr) {
+         block_log::construct_index(block_file_name, index_file_name);
+      
+         ind_in = FC_FOPEN(index_file_name.generic_string().c_str(), "rb");
+         EOS_ASSERT( ind_in != nullptr, block_log_not_found, "cannot open file ${file} after it is constructed", ("file",index_file_name.string()) );
+      }
+
+      if (is_head_block_valid()) {
+         // It is possible index file and
+         // block file do not agree with each other, make them match.
+         make_indexfile_agree_with_blockfile();
+      } else {
+         // Head block is bad. Find the last good block in block file,
+         // and make block file and index file agree with each other
+         trim_to_last_good_block();
+      }
+   }
+
+   bool trim_data::is_head_block_valid() {
+      // Go to the location storing head block's starting position
+      auto status = fseek(blk_in, -sizeof(uint64_t), SEEK_END);
+      EOS_ASSERT( status == 0, block_log_exception, "cannot seek to ${file} ${pos} from end of file", ("file", block_file_name.string())("pos", sizeof(uint64_t)) );
+
+      // Read head block's starting position
+      uint64_t head_pos;
+      auto size = fread((void*)&head_pos, sizeof(uint64_t), 1, blk_in);
+      EOS_ASSERT( size == 1, block_log_exception, "${file} read fails", ("file", block_file_name.string()) );
+
+      // If head block is serializable, it is considered valid
+      return (is_block_serializable(head_pos));
+   }
+
+   bool trim_data::is_block_serializable(uint64_t pos) {
+      if (pos > block_file_size) {
+         ilog( "Returning false because pos ${pos} greater than ${block_file_size}", ("pos", pos) ("block_file_size", block_file_size) );
+         return false;
+      }
+
+      auto status = fseek(blk_in, pos, SEEK_SET);
+      if (status != 0) {
+         ilog( "Could not seek past ${bytes} bytes in Block log file at '${blocks_log}'. Returned status: ${status}",
+               ("bytes", pos)("blocks_log", block_file_name)("status",status) );
+         return false;
+      }
+
+      detail::fileptr_datastream ds(blk_in, block_file_name.string());
+
+      try {
+         if (version >= pruned_transaction_version) {
+            detail::log_entry_v4 entry;
+            fc::raw::unpack(ds, entry.meta.size);
```

Comment:
You didn't actually deserialize the block here.

Reply:
I only check if the header can be deserializable or not for a quick check if the block is corrupt or not. Later I also compare block IDs. Those do not cover all cases of corruption; in my testing, they caught all.

Reply:
Are you aware of any better ways to detect block corruption? Thanks.

Reply:
You should use log_entry instead of log_entry_v4 to deserialize the data.

Reply:
Thanks for spotting this! 
I made a mistake when merging with latest code. The corrected code now looks like:
      detail::fileptr_datastream ds(blk_in, block_file_name.string());
      try {
        detail::log_entry entry;
        if (version < pruned_transaction_version) {
           entry.emplace<signed_block_v0>();
        }
        eosio::chain::detail::unpack(ds, entry);
      } catch (...) {
         ilog("Failed to desearialize, returning false");
         return false;  // If we cannot unpack (desearialize), it is certain the block is corrupt
      }

I have also added 
size_t get_stream_pos(fileptr_datastream& ds) { return ds.ftell(); };
void   skip_streamm_pos(fileptr_datastream& ds, size_t n) { ds.skip(n); }
so that unpack works for fileptr_datastream.

I have also changed the function name to is_block_deserializable